src/lib/comms/sol-http-common.c

```diff
@@ -74,52 +130,645 @@ sol_http_param_free(struct sol_http_param *params)
+        SOL_INT_CHECK_GOTO(r, < 0, exit);
     }
 
     r = sol_buffer_ensure_nul_byte(&buf);
-    SOL_INT_CHECK_GOTO(r, < 0, end);
+    SOL_INT_CHECK_GOTO(r, < 0, exit);
+
+    *url_out = sol_buffer_steal(&buf, NULL);
+exit:
+    sol_buffer_fini(&buf);
+    return r;
+}
+
+SOL_API int
+sol_http_create_simple_uri(char **uri, const struct sol_str_slice base_url,
+    const struct sol_http_param *params)
+{
+    struct sol_buffer buf;
+    int r;
+
+    SOL_NULL_CHECK(uri, -EINVAL);
+    if (base_url.len == 0) {
+        SOL_WRN("base_url is empty!");
+        return -EINVAL;
+    }
+
+    sol_buffer_init(&buf);
+
+    r = sol_buffer_append_slice(&buf, base_url);
+    SOL_INT_CHECK(r, < 0, r);
 
-    *escaped = sol_buffer_steal(&buf, NULL);
+    if (params && params->params.len) {
+        r = sol_buffer_append_slice(&buf, sol_str_slice_from_str("/?"));
+        SOL_INT_CHECK_GOTO(r, < 0, exit);
+        r = sol_http_encode_params(&buf, SOL_HTTP_PARAM_QUERY_PARAM,
+            params);
+        SOL_INT_CHECK_GOTO(r, < 0, exit);
+    }
 
-end:
+    *uri = sol_buffer_steal(&buf, NULL);
+exit:
     sol_buffer_fini(&buf);
     return r;
 }
+
+static void
+_update_partial_uri(struct sol_str_slice *partial_uri, size_t offset)
+{
+    partial_uri->data = partial_uri->data + offset;
+    partial_uri->len = partial_uri->len - offset;
+}
+
+static int
+_get_scheme(const struct sol_str_slice uri,
+    struct sol_str_slice *next,
+    struct sol_str_slice *scheme)
+{
+    char *sep;
+
+    scheme->data = uri.data;
+
+    sep = memchr(uri.data, ':', uri.len);
+
+    if (!sep) {
+        SOL_WRN("Could not find the scheme seperator (:) at URI:%.*s",
+            SOL_STR_SLICE_PRINT(uri));
+        return -EINVAL;
+    }
+
+    scheme->len = sep - scheme->data;
+
+    if (!scheme->len) {
+        SOL_WRN("Empty scheme. URI: %.*s", SOL_STR_SLICE_PRINT(uri));
+        return -EINVAL;
+    }
+
+    SOL_DBG("URI Scheme: %.*s", SOL_STR_SLICE_PRINT(*scheme));
+    _update_partial_uri(next, scheme->len + 1);
+
+    return 0;
+}
+
+static char *
+_get_first_sep(char *sep1, char *sep2)
+{
+    if (!sep1)
+        return sep2;
+    if (!sep2)
+        return sep1;
+    return sol_min(sep1, sep2);
+}
+
+static int
+_get_authority(const struct sol_str_slice partial_uri,
+    const struct sol_str_slice full_uri,
+    struct sol_str_slice *next,
+    struct sol_str_slice *host,
+    struct sol_str_slice *user,
+    struct sol_str_slice *pass,
+    int *port)
+{
+
+    size_t offset = 2;
+    struct sol_str_slice port_slice = SOL_STR_SLICE_EMPTY;
+    char *sep, *port_sep, *fragment_sep, *path_sep, *query_sep;
+    struct sol_str_slice auth = partial_uri;
+
+    if (!auth.len) {
+        SOL_WRN("Empty authority URI: %.*s", SOL_STR_SLICE_PRINT(full_uri));
+        return -EINVAL;
+    }
+
+    //Not an URL
+    if (strncmp(auth.data, "//", 2))
+        return 0;
+
+    *port = -1;
+    auth.data += 2;
+    auth.len -= 2;
+
+    sep = memchr(auth.data, '@', auth.len);
```

Comment:
this looks easier, but will do useless work and is not that easier. It's better to just do a single for() until the authority terminates and while doing so, set pointers to parts:

```c
   const char *itr = auth.data;
   const char *itr_end = itr + auth_data.len;
   struct sol_str_slice user = SOL_STR_SLICE_STR(NULL, 0); /* so we get NULL instead of "" */
   struct sol_str_slice pass = SOL_STR_SLICE_STR(NULL, 0);
   struct sol_str_slice host = SOL_STR_SLICE_STR(NULL, 0);
   struct sol_str_slice port = SOL_STR_SLICE_STR(NULL, 0);
   const char *last = auth.data;
   for (; itr < itr_end; itr++) {
       const char c = *itr;
       if (c == '/') {
           auth.len = itr - auth.data;
           if (host.len == 0) {
               host.len = itr - host.data;
          } else if (port.data) {
               port.len = itr - port.data;
          }
          break;
       } else if (c == ':') {
           /* could be host:port or user:pass, store in host:port and move later if '@' appears */
           if (host.data) {
               return -EINVAL; // already got a host:port (or user:port)
           }
           host.data = last;
           host.len = itr - host.data;
           last = itr + 1;
           port.data = last;
           port.len = 0;
       } else if (c == '@') {
           if (user.len) {
               return -EINVAL; // already got an user
           } else {
               user = host;
               pass = port;
               host = SOL_STR_SLICE_STR(NULL, 0);
               port = SOL_STR_SLICE_STR(NULL, 0);
               last = itr + 1;
           }
       }
   }
```

I didn't test it, but should be something like that.